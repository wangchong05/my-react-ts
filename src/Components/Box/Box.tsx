import React from "react";
import styled, { StyledProps } from "styled-components";
import axios from "axios";
import List from "../List/List";

export interface BoxProps extends StyledProps<any> {
  label?: string;
  // list: Array<object>;
  name?: string;
  age?: number;
}

const arr = [
  {
    id: 1,
    name: 'wang',
    age: 15
  }, {
    id: 2,
    name: 'wang2',
    age: 23
  }, {
    id: 3,
    name: 'wang3',
    age: 44,
    done: true
  }, {
    id: 4,
    name: 'wang4',
    age: 42
  }, {
    id: 5,
    name: 'wang5',
    age: 8
  },
]

const BoxWrapper: React.FC<BoxProps> = styled.div<BoxProps>`
  padding: 10px;
  border: 1px solid gray;
`;

const getData = () => {
  axios.get('https://jsonplaceholder.typicode.com/posts').then(res => {
    console.log(res.data)
    const list: Array<object> = res.data;
    return <List list={list}/>
  })
}

const getData2 = (list: Array<object>) => {
  return (
    list.map((item: object, index: any) => {
      const { name, age }: BoxProps = item;
      return (
        <div key={index} style={{'color': 'gold'}}>
          <span>{index}</span>
          <span>_</span>
          <span>{name}</span>
          <span>_</span>
          <span>{age}</span>
          <hr/>
        </div>
      )
    })
  );
}

const getData3 = () => {
  return <List list={arr}/>
  
}

const Box = ({ label, children }: BoxProps) => {
  getData()

  // return <div data-testid='box'>{ label }</div>;

  return (
    <div>
      {/* <List list={arr}/> */}

      {/* <div>{getData()}</div> */}
      {/* 不能将类型“void”分配给类型“ReactNode”。ts(2322)
index.d.ts(1346, 9): 所需类型来自属性 "children"，在此处的 "DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>" 类型上声明该属性
const getData: () => void */}
      <div>{getData3()}</div>

      {/* {getData2(arr)} */}
      <BoxWrapper data-testid='box'>
        { label }
      </BoxWrapper>
    </div>
  );
}

export default Box;