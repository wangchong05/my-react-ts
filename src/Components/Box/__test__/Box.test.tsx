import React from "react";
// import { ReactDOM } from "react";
import { render, screen } from '@testing-library/react';
import ReactDom from "react-dom";
// import "jest-dom/extend-expect";
import renderer from "react-test-renderer";
import Box from "./../Box"

// 首先看该组件会不会渲染
test("renders without crashing", ()=> {// 渲染而不崩溃
  const div = document.createElement("div")// 创建一个div，然后尝试渲染
  ReactDom.render(<Box></Box>, div)// 把Box渲染在刚才创建的div上
})

// 正确渲染了Box
test("renders Box correctly", ()=> {
  const { getByTestId } = render(<Box label={"box test content"}></Box>)// 需要使用render函数，这次在Box里传递label。需要使用数据测试访问Box元素
  // 因此函数返回一个getByTestId的函数
  expect(getByTestId('box')).toHaveTextContent("box test content");// 用期望将getByTestId做包装，通过数据测试ID获得这个节点
})

// 如果想通过传递不同的标签来渲染其他内容，比如，想通过保存，将其保存在内部
test("renders Box correctly 02", ()=> {
  const { getByTestId } = render(<Box label={"save"}></Box>)
  expect(getByTestId('box')).toHaveTextContent("save");
})

// 
test("matchs snapshot", ()=> {
  const tree = renderer.create(<Box label={"save"}></Box>).toJSON();
  // 创建名为save的标签，将其转换为JSON，所以这会将其转换为虚拟DOM对象
  // 这个创建出来是个树的形式，已经将他保存到树里面，接下来期望树匹配快照
  expect(tree).toMatchSnapshot();
})

// 
test("matchs snapshot 02", ()=> {
  const tree = renderer.create(<Box label={"save 02"}></Box>).toJSON();
  expect(tree).toMatchSnapshot();
})