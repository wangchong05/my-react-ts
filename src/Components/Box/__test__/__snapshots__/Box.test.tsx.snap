// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchs snapshot 02 1`] = `
<div>
  <div
    className="sc-bdnxRM hhtRwE"
    data-testid="box"
  >
    save 02
  </div>
</div>
`;

exports[`matchs snapshot 1`] = `
<div>
  <div
    className="sc-bdnxRM hhtRwE"
    data-testid="box"
  >
    save
  </div>
</div>
`;
